### DISCUSSION TITLE
Add partstat to CalendarEvent

### DISCUSSION THREAD

**Discussion by joelhawksley** [home-assistant/architecture#1137] [Entity Models] [answered]
## Context

When working with calendar events, it can be useful to filter by response status. For example, one might want to hide events that have been declined. 

## Proposal

Add a `partstat` field to CalendarEvent per [rfc5545](https://icalendar.org/iCalendar-RFC-5545/3-2-12-participation-status.html). 

## Impact

As this change adds a field, I do not believe it will cause any breakages.

## Proof of concept

https://github.com/home-assistant/core/pull/125708

**Comment by allenporter** [comment]
PARTSTAT is not a top level rfc5545 component property.  It is a property parameter of another property, Attendee. As a result, this feels like it should be https://github.com/home-assistant/architecture/discussions/1099

Would be good to be specific about which calendar integrations you plan to update to use this and how it will work.

**Reply by joelhawksley** [reply]
@allenporter I use Google Calendar personally. I would imagine setting the field to be one of the values in the partstat spec as enumerated in my PoC pull request.

**Reply by allenporter** [reply]
@joelhawksley let me clarify the intent behind my questions further since there is more here i'm looking for:
(1) Changes to the entity model need to be more general and not for a single use case of a single integration.  This needs to consider more than just a single integration, for example how this could work in say local calendar, caldav, etc.
(2) We need more of a spec of what this means, and you didn't acknowledge what I said about the specification of PARTSTAT, that  PARTSTAT on it's own is not a thing, and that it is a property of an attendee and there is already ad discussion on this. It needs to have well defined behavior that can be explained here about what it means.  I would assume what you mean by this is that it's the value for the attendee who is the calendar owner, but you have not spelled that out in detail, what the rules are, how corner cases are handled, what are the implications for attendees, and to (1) above how this would work in other integrations as well. The naming `partstat` on its own 

As I mentioned in https://github.com/home-assistant/architecture/discussions/1099#discussioncomment-9758697 I think it would be way simpler to just make a breaking change to the behavior of google calendar since it probably matches what users want. As a codeowner for google calendar integration, i'd support this.

**Reply by joelhawksley** [reply]
@allenporter thank you for the clarification! This is my first foray into contributing to HA ❤️ 

I had assumed that since PARTSTAT is part of the RFC mentioned in the base calendar integration source that it would be acceptable to add it/an attendees list to CalendarEntity and assume that any other calendar integration would generally be able to conform to that specification. For example, at a cursory glance I do see PARTSTAT mentioned in CalDAV: https://datatracker.ietf.org/doc/html/rfc4791.

I agree with the argument for making the breaking change to exclude events marked as `declined` by the integration owner (which Google labels `self` in the attendee list API response). Do you imagine eventually doing the same for other calendar integrations?

**Reply by allenporter** [reply]
I don't know if this applies to them or not. Caldav maybe but I don't know how you identify the owner of the calendar?

To be clear: generally it's great to add things from the RFC! The problem is this is that partstat is attendee property and so it needs a story about how that factors in.

**Reply by joelhawksley** [reply]
@allenporter that's a great point. There is no concept of an owner when loading a CalDAV URL. 

So, in that case, maybe we would add a `self_partstat` field that is only populated for calendar integrations that define the viewer of the calendar events?

**Reply by allenporter** [reply]
@joelhawksley i think that's the question as its not clear how all that could work and what integrations it could work on.

This is why I go back to: If the goal is to filter declined events from google calendar that are declined then maybe the desire is to just change google calendar

**Reply by joelhawksley** [reply]
Ultimately, I think either way is fine with me. I'd be happy to help implement whichever path makes the most sense to y'all. How can I help move this forward?

**Reply by allenporter** [reply]
As i mentioned above: I think it would be way simpler to just make a breaking change to the behavior of google calendar since it probably matches what users want. As a codeowner for google calendar integration, i'd support this.

**Reply by joelhawksley** [reply]
Ok, would you be 👍🏻 to me trying to write a PR to `core` to do so?

**Reply by allenporter** [reply, marked as answer]
Yes, happy for you to send a PR.

**Comment by sfdye** [comment]
Hi folks @joelhawksley @allenporter Happy new year! Sorry I could not comment on the https://github.com/home-assistant/core/pull/128900 because the discussion is closed there.

I notice this new feature is included 2024.11 ([changlog](https://www.home-assistant.io/blog/2024/11/06/release-202411/)), however it's not working for me, because the attendee data is corrupted (always defaulting to `display_name=None,  response_status=ResponseStatus.NEEDS_ACTION`. 

I've mentioned this issue in my other [PR](https://github.com/home-assistant/core/pull/119457#issuecomment-2183632091) as well.

**Reply by allenporter** [reply]
We should clarify the state of this discussion: The proposal that was accepted was to never include declined events in the calendar.

You seem to be discussing something unrelated to the architecture issue, which is behavior of a specific client library for a specific integration -- so this isn't the right place for it, let's move this somewhere else.

My impression is this is a continued discussion of the topic you previously filed <https://github.com/allenporter/gcal_sync/issues/545> where I gave you some guidance on how to investigate further.

**Reply by sfdye** [reply]
Oops, I totally forgot about that issue. Let's move the discussion over there.
